File: Java/String/008 Rotate String.java
/**
 * Given a string and an offset, rotate string by offset. (rotate from left to right)
 * Example
 * Given "abcdefg".
 * offset=0 => "abcdefg"
 * offset=1 => "gabcdef"
 * offset=2 => "fgabcde"
 * offset=3 => "efgabcd"
 */

public class Solution {
    /**
     * @param str: an array of char
     * @param offset: an integer
     * @return: nothing
     */
    public void rotateString(char[] str, int offset) {
        // write your code here
        if (offset == 0 || str == null || str.length == 0) {
            return;
        }
        offset = offset % str.length;
        swap(str, 0, str.length - 1);
        swap(str, 0, offset - 1);
        swap(str, offset, str.length - 1);
        return;
    }
    private void swap(char[] str, int start, int end) {
        while (start < end) {
            char temp = str[start];
            str[start] = str[end];
            str[end] = temp;
            start++;
            end--;
        }
        return;
    }
}

File: C++/String/008 Rotate String.cpp
/**
 * Given a string and an offset, rotate string by offset. (rotate from left to right)
 * Example
 * Given "abcdefg".
 * offset=0 => "abcdefg"
 * offset=1 => "gabcdef"
 * offset=2 => "fgabcde"
 * offset=3 => "efgabcd"
 */

class Solution {
public:
  /**
     * param A: A string
     * param offset: Rotate string with offset.
     * return: Rotated string.
     */
    string rotateString(string A, int offset) {
        // wirte your code here
        string s;
        int sz = A.length();
        if(sz == 0)
            return A;
        offset = offset%sz;
        for(int i = sz- offset; i<sz; i++){
            s.push_back(A[i]);
        }
        for(int i =0; i<sz-offset; i++){
            s.push_back(A[i]);
        }
        return s;
    }
};